muting: Update the settings/muted-topics page UI.

ryanreh99 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This is a continuation of [PR#13702](https://github.com/zulip/zulip/pull/13702)
This updates the `#settings/muted-topics` page by refactoring `set_up_muted_topics_ui` to use `list_render`. And also include a new field `date_muted`.

**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![topic_mutes_ui](https://user-images.githubusercontent.com/40122794/74605303-f029da00-50ec-11ea-837a-29452aca5ac2.gif)
EDIT: Updated the gif as the previous gif contained a bug (if a topic gets unmuted, then the table rerenders with the streams sorted instead of the current sort configuration).

ryanreh99 commented on zerver/tests/test_events.py:
```diff
@@ -1525,7 +1525,7 @@ def test_default_streams_events_guest(self) -> None:
         muted_topics_checker = self.check_events_dict([
             ('type', equals('muted_topics')),
-            ('muted_topics', check_list(check_list(check_string, 2))),
+            ('muted_topics', check_list(check_list(None))),
```
`None`, because the function can only validate if all items are of the same type.
There is a todo for `realm_filters` in the same file: `TODO: validate tuples in the list`

timabbott replied:
Just to make sure I understand this, we're changing a 2-element list to a 3-element list?  I think we should at least preserve the `, 2` -> `, 3` part of the check here.  

In any case, this is a wire format API change that we need to run by the mobile folks.  @gnprice how does the mobile app parse the `muted_topics` event type?  Would changing the format of that to add an extra element to the list be a problem?

ryanreh99 replied:
Connected zulip-mobile with dev server, and everything seems normal, the topics gets displayed as it should and I am able to mute topics. (Attached the testing below)

![zulip_mobile_debug](https://user-images.githubusercontent.com/40122794/74605958-7bf23500-50f2-11ea-8c57-8249b6415b89.gif)

ryanreh99 commented on static/js/muting_ui.js:
```diff
@@ -113,7 +112,7 @@ exports.mute = function (stream_id, topic) {
         undo_button_text: i18n.t("Unmute"),
     });
-    exports.set_up_muted_topics_ui(muting.get_muted_topics());
+    // exports.set_up_muted_topics_ui(muting.get_muted_topics());
```
With this line I cannot mute a topic unless the settings page has been opened first.
`const muted_topics_table = $("#muted_topics_table").expectOne();` this is the line causing the error (attached image below).

I think this is happening because the `#muted-topic-settings` html element hasn't gotten loaded yet, so jquery can't select it.
idk, if we require this line as all topic mutes are done from the homepage and after commenting it out, the above bug seems to be resolved.
And it doesn't seem this line is used for real time sync, as on opening 2 tabs, if a topic gets muted and if the other tabs is on the `#settings/muted-topics`page, the table gets updated only after reloading.

im not sure how to proceed with this, could i get some help?
![Screenshot from 2020-02-06 18-49-52](https://user-images.githubusercontent.com/40122794/73948019-67f24a80-491e-11ea-9577-6515e1dea7bb.png)

timabbott replied:
@pragatiagrawal31 @aero31aero @hackerkid does one of you have time to help with this?

aero31aero replied:
I don't know why it was originally written this way but we do not need to call the `set_up_muted_topics_ui` function here; we already call it when we open the settings UI (see `static/settings_muted.js`). We should remove these two calls:

```diff
diff --git static/js/muting_ui.js static/js/muting_ui.js
index b473303624..bb8a51480f 100644
--- static/js/muting_ui.js
+++ static/js/muting_ui.js
@@ -112,7 +112,6 @@ exports.mute = function (stream_id, topic) {
         title_text: i18n.t("Topic muted"),
         undo_button_text: i18n.t("Unmute"),
     });
-    // exports.set_up_muted_topics_ui(muting.get_muted_topics());
 };
 
 exports.unmute = function (stream_id, topic) {
@@ -124,7 +123,6 @@ exports.unmute = function (stream_id, topic) {
     unread_ui.update_unread_counts();
     exports.rerender();
     exports.persist_unmute(stream_id, topic);
-    exports.set_up_muted_topics_ui(muting.get_muted_topics());
     feedback_widget.dismiss();
 };
```

aero31aero replied:
I wonder if there are other similar bugs that have gone by undetected because we do not use `expectOne` or similar logic everywhere.

timabbott commented on zerver/lib/topic_mutes.py:
```diff
@@ -20,17 +20,19 @@
-    return [
-        [row['stream__name'], row['topic_name']]
+    result = [
+        [row['stream__name'], row['topic_name'], datetime.datetime.timestamp(row['date_muted'])]
```
This should use the helpers from `zerver/lib/timestamp.py`.

timabbott commented:
@ryanreh99 let me know when you have this updated; I think it's not far from being mergable.

ryanreh99 commented:
hello @timabbott, I updated the PR and the gif.
Also left a comment about testing the api change on zulip-mobile above.

timabbott commented:
@ryanreh99 can your rebase this?  I think it's pretty close to mergable.

ryanreh99 commented:
@timabbott ping.

timabbott commented:
I merged the first commit.  The second needs a rebase and fixing for the recent `list_render` API changs of @showell's that we just merged, but otherwise lgtm.

timabbott commented:
(Let me know when you have a chance to rebase this)

ray-kraesig commented:
Shouldn't `List[Union[str, float]]` be `Tuple[str, str, float]`?

ryanreh99 commented:
> Shouldn't `List[Union[str, float]]` be `Tuple[str, str, float]`?

Thanks! This does look more accurate.

```Vim script
zerver/lib/topic_mutes.py:32: error: List comprehension has incompatible type List[List[Any]]; expected List[Tuple[str, str, float]] 
```

Although, I am not sure about this error. I'll look it up.

ray-kraesig commented:
It looks like mypy can't tell that the list comprehension is supposed to be generating a Tuple.

I think you can put `  # type: Tuple[str, str, Float]` immediately after that subexpression to resolve it; if not, you'll probably have to create an auxiliary function with the appropriate type.

ryanreh99 commented:
> I think you can put # type: Tuple[str, str, Float] immediately after that subexpression to resolve it; if not, you'll probably have to create an auxiliary function with the appropriate type.

```Vim script
error: misplaced type annotation
```
I went with the second method, as I couldn't resolve the above error.

I think github/python/mypy/5068 is the reason for the incompatible list comprehension type and should be included in `mutes_info` function's comment but im not too sure. (So didn't want to tag this link on the other issue's page)

ryanreh99 commented:
@timabbott, Rebased this and made the following changes:-
• a minor change in the `muting.add_muted_topic ` to remove the comparision with `undefined`.
• fixed the real time sync bug. (left it un-squashed for now)
• added a commit about the type annotation issue as mentioned by @ray-kraesig. (added more details in the comment above.)

(Sorry about the delay)

timabbott commented on zerver/lib/validator.py:
```diff
@@ -189,7 +189,7 @@ def check_list(sub_validator: Optional[Validator], length: Optional[int]=None) -
     @set_type_structure(type_structure)
     def f(var_name: str, val: object) -> Optional[str]:
-        if not isinstance(val, list):
+        if not (isinstance(val, list) or isinstance(val, tuple)):
```
What's the story with this change?  It seems unlikely to be correct in some sense; but I guess the issue is there's no difference between a tuple and a list in JSON?

ryanreh99 replied:
the schema check test mentioned in [this comment above](https://github.com/zulip/zulip/pull/13833#discussion_r375878827), was failiing.

showell replied:
I would really want to avoid modifying `check_list` just to make it work for the somewhat unique muting use case.  We should get away from tuples on that API and just send a proper struct in.  Or, if we do keep tuples, just use a custom validator, so we can type-check the individual fields.  But we don't want any of the muting sloppiness to leak into other code.

timabbott commented:
I merged the UI commits; the typing commit I still have questions about.  Thanks for cleaning up that UI element @ryanreh99!

timabbott commented:
I'm not sure we want to change `check_list`, which we also use for user input validation, to accept tuples.

Is there actually a difference between a tuple and a list in the JSON we send via the API?  It's not clear to me that typing this as a tuple, vs. a 3-element list, is actually helpful.

ray-kraesig commented:
> Is there actually a difference between a tuple and a list in the JSON we send via the API? It's not clear to me that typing this as a tuple, vs. a 3-element list, is actually helpful.

There's no difference in the serialization, but the schema may differ. Consumers would like to know that it's specifically a `[string, string, number]` rather than potentially a `[number, string, string]`.

I wouldn't think modifying `check_list` is the right approach either, though; at first glance, this looks like it wants to use a `check_tuple` validator which takes a list – or tuple – of other validators.

ryanreh99 commented:
As there is already a `TODO: validate tuples in the list`, I'll start by creating a `check_tuple` validator.

Since mypy cant annotate a list of multiple types, should we use a different data structure instead of tuple?

timabbott commented:
@ryanreh99 I think having mypy treat it as a `List[Union[str, int]]` should be good enough.  There's not enough code touching this to be worth doing something more fancy.

timabbott commented:
@ryanreh99 ping on this; it'd be nice to close out the bit remaining in this PR.

ryanreh99 commented:
@timabbott Sorry, could you mention what should be done?
> I think having mypy treat it as a List[Union[str, int]] should be good enough.

In the merged commit [9340cd1](https://github.com/zulip/zulip/commit/9340cd1a0b7af7af9b83a622e37d02f1a2340609) it is `List[Union[str, float]]`

timabbott commented:
OK, so should we just close this with nothing further worth doing?
